state/kvstore.go

```diff
@@ -32,107 +73,80 @@ func (mkv *MemKVStore) Get(key []byte) (value []byte) {
 	return mkv.m[string(key)]
 }
 
-//----------------------------------------
-
-// A Cache that enforces deterministic sync order.
-type KVCache struct {
-	store    KVStore
-	cache    map[string]kvCacheValue
-	keys     *list.List
-	logging  bool
-	logLines []string
+func (mkv *MemKVStore) Has(key []byte) (has bool) {
```

Comment:
You are correct, but since the whole blockchain is about serializing transactions, this doesn't apply.  You can have two caches over one store (check and deliver), but in the end one will be discarded and there are never concurrent writes while you are reading.